### ISSUE TITLE
build doesn't typecheck

### ISSUE THREAD

**Issue by ocramz** [haskell/c2hs#118]
Install command: 
cabal install haskell-mpi --extra-lib-dirs=/opt/openmpi-1.5.3/lib/ --extra-include-dirs=/opt/openmpi-1.5.3/include/

[1 of 6] Compiling C2HS             ( src/C2HS.hs, dist/dist-sandbox-63b7207b/build/C2HS.o )

src/C2HS.hs:165:31:
    Could not deduce (Num b) arising from the literal `0'
    from the context (Enum a, Bits b)
      bound by the type signature for
                 combineBitMasks :: (Enum a, Bits b) => [a] -> b
      at src/C2HS.hs:164:20-47
    Possible fix:
      add (Num b) to the context of
        the type signature for
          combineBitMasks :: (Enum a, Bits b) => [a] -> b
    In the second argument of `foldl', namely `0'
    In the first argument of `(.)', namely `foldl (.|.) 0'
    In the expression: foldl (.|.) 0 . map (fromIntegral . fromEnum)

src/C2HS.hs:171:39:
    Could not deduce (Num a) arising from a use of `fromIntegral'
    from the context (Bits a, Enum b)
      bound by the type signature for
                 containsBitMask :: (Bits a, Enum b) => a -> b -> Bool
      at src/C2HS.hs:170:20-53
    Possible fix:
      add (Num a) to the context of
        the type signature for
          containsBitMask :: (Bits a, Enum b) => a -> b -> Bool
    In the first argument of `(.)', namely `fromIntegral'
    In the expression: fromIntegral . fromEnum
    In the expression: fromIntegral . fromEnum $ bm

follows my ghc --info: 
 
 [("Project name","The Glorious Glasgow Haskell Compilation System")
 ,("GCC extra via C opts"," -fwrapv")
 ,("C compiler command","/usr/bin/gcc")
 ,("C compiler flags"," -fno-stack-protector")
 ,("C compiler link flags","")
 ,("ld command","/usr/bin/ld")
 ,("ld flags","")
 ,("ld supports compact unwind","YES")
 ,("ld supports build-id","YES")
 ,("ld supports filelist","NO")
 ,("ld is GNU ld","YES")
 ,("ar command","/usr/bin/ar")
 ,("ar flags","q")
 ,("ar supports at file","YES")
 ,("touch command","touch")
 ,("dllwrap command","/bin/false")
 ,("windres command","/bin/false")
 ,("libtool command","libtool")
 ,("perl command","/usr/bin/perl")
 ,("target os","OSLinux")
 ,("target arch","ArchX86_64")
 ,("target word size","8")
 ,("target has GNU nonexec stack","True")
 ,("target has .ident directive","True")
 ,("target has subsections via symbols","False")
 ,("Unregisterised","NO")
 ,("LLVM llc command","llc")
 ,("LLVM opt command","opt")
 ,("Project version","7.8.2")
 ,("Booter version","7.6.3")
 ,("Stage","2")
 ,("Build platform","x86_64-unknown-linux")
 ,("Host platform","x86_64-unknown-linux")
 ,("Target platform","x86_64-unknown-linux")
 ,("Have interpreter","YES")
 ,("Object splitting supported","YES")
 ,("Have native code generator","YES")
 ,("Support SMP","YES")
 ,("Tables next to code","YES")
 ,("RTS ways","l debug thr thr_debug thr_l thr_p dyn debug_dyn thr_dyn thr_debug_dyn l_dyn thr_l_dyn")
 ,("Support dynamic-too","YES")
 ,("Support parallel --make","YES")
 ,("Dynamic by default","NO")
 ,("GHC Dynamic","YES")
 ,("Leading underscore","NO")
 ,("Debug on","False")
 ,("LibDir","/home/mzocca/bin/lib/ghc-7.8.2")
 ,("Global Package DB","/home/ocramz/bin/lib/ghc-7.8.2/package.conf.d")
 ]

**Comment by ian-ross** [comment]
@ocramz Not sure I would call this a C2HS bug.  The `haskell-mpi` package hasn't been updated since 2012 -- since then, C2HS has changed considerably: in particular, the `C2HS.hs` file is no longer used at all, and the syntax for function hook definitions has changed so that much of the usage in this package is not valid with recent versions of C2HS (in particular a lot of the stuff in `Control.Parallel.MPI.Internal` is now incompatible with the way that C2HS function hooks work).  You could try using an old (2012-vintage) version of C2HS, but I still think you'd have trouble with using that C2HS.hs with more recent GHC versions.

From a quick skim through the C2HS code, it looks as though it shouldn't be too hard to update this for current versions of C2HS.  If you think that would be useful, I can do it (or you could do it and I could cheer from the sidelines...).  The maintainer of the `haskell-mpi` package is @bjpop who seems pretty active, so hopefully he'd accept a PR to bring the package up to date.

Let me know what you think: if you really want to use this package, I'll update it (I might have some interest in using MPI with Haskell at some point in the future as well).

**Comment by ocramz** [comment]
Hi Ian, @ian-ross , thank you for the quick reply.
I have already tried with combinations of vintage c2hs/haskell-mpi, to no avail.
I haven't had any experience in FFI bindings yet, but perhaps if you instructed me on what to change and why, I might try my hand.. 
In the meanwhile; yes, I do mean to use this package, since my uni has a big fat compute cluster and it would be a shame to let it idle.

**Comment by ian-ross** [comment]
@ocramz OK, I'll make the necessary changes some time in the next week.  It's probably one 1-2 hours of work and if you've not done much FFI stuff before, it will be quicker if I do it.  I'll let you know when it's working and you can give it a go before I send a PR to the maintainer of `haskell-mpi`.

**Comment by ian-ross** [comment]
@ocramz Try this (obviously change the OpenMPI paths for whatever you have set up; you might need an include file path as well):

```
git clone git@github.com:ian-ross/haskell-testrunner.git
git clone git@github.com:ian-ross/haskell-mpi.git
cd haskell-mpi
git checkout latest-c2hs
cabal sandbox init
cabal sandbox add-source ../haskell-testrunner
cabal install --extra-lib-dirs=/usr/lib/openmpi -ftest --only-dep
cabal configure --extra-lib-dirs=/usr/lib/openmpi -ftest
cabal build
mpirun -np 2 ./dist/build/haskell-mpi-testsuite/haskell-mpi-testsuite
```

The changes for the latest C2HS are on the `latest-c2hs` branch of my `haskell-mpi` repo.  The `testrunner` stuff is needed because `haskell-mpi` uses an old test management package for running its test suite and there have been some changes in QuickCheck since that was last released.

This all works with GHC 7.8.4, cabal-install 1.22.0.0 and OpenMPI 1.8.4 on my Arch Linux box.  There are some failures in the `haskell-mpi` test suite that I've not investigated, but most of the tests seem to pass.

Let me know if this works and if it does, I'll pass a PR to the maintainer of `haskell-mpi`.

**Comment by ocramz** [comment]
Hi Ian, thanks a lot for this work; however installation fails at the `cabal configure` step, with a 
Configuring haskell-mpi-1.3.0...
cabal: The program 'c2hs' is required but it could not be found.

**Comment by ian-ross** [comment]
Ah sorry, I assumed you already had C2HS installed.  You need to have C2HS somewhere on your path: Cabal doesn't manage build tools dependencies in the way that it manages package dependencies.  Just to `cabal install c2hs` outside a sandbox and make sure you have `~/.cabal/bin` on your path.

**Comment by ocramz** [comment]
Installed c2hs globally, everything works now. However `mpirun` complains as follows:

$ mpirun -np 2 ./dist/build/haskell-mpi-testsuite/haskell-mpi-testsuite
Need at least two processes to run the tests.
Typical command line could look like this:
'mpirun -np 2 bindings-mpi-testsuite 1>sender.log 2>receiver.log'

Need at least two processes to run the tests.
Typical command line could look like this:
'mpirun -np 2 bindings-mpi-testsuite 1>sender.log 2>receiver.log'

**Comment by ian-ross** [comment]
That's odd.  Is your MPI installation OK?  I.e. does it work with a simple C test program?

**Comment by ocramz** [comment]
Yes, I've just tested it with this: http://techtinkering.com/2009/12/02/setting-up-a-beowulf-cluster-using-open-mpi-on-linux/ , processes send wait and receive as they should. What could the problem be?

**Comment by ian-ross** [comment]
I really don't know what could be going on there.  At that point, all the test program has done is call `MPI_Init` and call the Haskell equivalent of `MPI_Comm_size` on the world communicator.  I don't know if the initialisation is failing and so you're getting something funny back as the communicator size.

What sort of system are you using?  Is it Linux?  Just a desktop machine (i.e. no funny cluster stuff)?  And OpenMPI, you say?  I'm pretty surprised that it's not working!

**Comment by ocramz** [comment]
I'm actually using a Torque/PBS cluster running CentOS. I already tried sending the `mpirun` command from within a PBS batch job, but the error message is the same ("Need at least two processes to run the tests")

**Comment by ian-ross** [comment]
OK, can you try doing it just on your desktop machine or some other single-node machine first, just to make sure that you're getting a good build of the `haskell-mpi` package and it's able to at least initialise the version of OpenMPI you're using?  Getting it working on a real cluster is a (quite big!) second step.

**Comment by ocramz** [comment]
It will take me some time, as I don't have a local MPI installation. In the meanwhile, I've just noticed that the error message mentions `bindings-mpi-testsuite`, whereas the binary we're calling is `haskell-mpi-testsuite`. Any relation?

**Comment by ian-ross** [comment]
I think that's just a typo in the message.  I don't have access to a cluster for testing right now, though I should do some time in the next couple of months.  It's obviously a bit difficult to guess what might be going wrong since there are so many variables.  I can't quite believe that the MPI initialization is failing completely without throwing an error.  Something a bit weird is going on.

**Comment by bjpop** [comment]
Hi ian-ross and ocramz. Thanks for looking into the build issues with haskell-mpi.
It appears as though some bit rotting has occurred.
I'll have a look at it myself and see if I can get it working again. I will happily accept pull requests if you make any improvements. Cheers!

**Comment by ian-ross** [comment]
@bjpop I'll send you a PR today or tomorrow -- I got it building and working with recent dependencies without too much trouble, at least on my desktop machine.

**Comment by gitfoxi** [comment]
I'm seeing this with the file `C2HS.hs` installed by `c2hs -l`. First there's a problem where it tries to `import Monad (liftM)` like it's 1995. After fixing that I get the type-checking errors as above. I will check out the latest branch as suggested by @ian-ross.

**Comment by ian-ross** [comment]
@gitfoxi This package is based on an obsolete version of C2HS that does not work with recent GHC versions.  I'm going to send @bjpop a PR today to allow `haskell-mpi` to build with recent GHC and C2HS.

**Comment by gitfoxi** [comment]
I see:

> module C2HS {-# DEPRECATED "The C2HS module should no longer be used." #-} (

So what should be used, generally for new projects?

**Comment by ian-ross** [comment]
The idea is that you shouldn't need that module because the functions it contained are either in the standard FFI modules (`Foreign.C.Types`, `Foreign.Ptr`, etc.) or they're baked into C2HS as default argument marshallers.  So, what you'll have instead of `import C2HS` is just a few imports for Haskell standard library FFI modules.

**Comment by ian-ross** [comment]
I'm going to close this as I think I've fixed the C2HS side of things -- why the `haskell-mpi` test suite doesn't seem to run for @ocramz isn't really a C2HS issue, I'd say.  I've sent a PR to @bjpop to update `haskell-mpi` to build with the latest C2HS.

**Comment by ocramz** [comment]
Thank you Ian!

**Comment by bjpop** [comment]
Sorry to re-open.

Thanks Ian for your work.

I've pulled Ian's changes and merged with the master branch of haskell-mpi. I've bumped the minor version number of the package.

All but one test case succeed on my machine. I'll look into the failed test later.

I can't upload the latest version to hackage because I'm not on the list of official maintainers there. I've asked to be added to that list. When that happens I'll make it available on hackage.
